### PR TITLE
ZEPPELIN-3411 Long running logic inside synchronized block in InterpreterSettingManager

### PR DESCRIPTION
**Pull request by jongyoul**
### What is this PR for?
Removing redundant synchronized code to avoid blocking other logics.

### What type of PR is it?
[Bug Fix]

### Todos
* [x] - Change synchronized block to read/write lock

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3411

### How should this be tested?
* Current tests should be passed

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by jongyoul** [comment]
This is not a critial issue when running normally. But in case of problemistic situation like restarting forcibly to stop long running job of SparkInterpreter, some logic hangs inside synchronized block.

**Comment by zjffdu** [comment]
Do you have jstack so that I can understand what the exact problem is

**Comment by jongyoul** [comment]
@zjffdu Let me make it

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java [hunk 1]
```diff
@@ -103,6 +104,7 @@
    */
   private final Map<String, InterpreterSetting> interpreterSettings =
       Maps.newConcurrentMap();
+  private final ReentrantReadWriteLock interpreterSettingsLock = new ReentrantReadWriteLock();
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Actually interpreterSettings is ConcurrentHashMap, I think we don't need extra lock here.

Comment (jongyoul) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Yes, I thought it and wanted to confirm it. Then, I'll remove all of the locks related to interpreterSettings in this class.

Comment (jongyoul) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Fixed

**Comment by jongyoul** [comment]
Removed all of the synchronized blocks from `InterpreterSettingManager` for `interpreterSettings`

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java [hunk 2]
```diff
@@ -940,17 +921,16 @@ public void close(String settingId) {
...
-        Thread t = new Thread() {
-          public void run() {
-            intpSetting.close();
-          }
-        };
-        t.start();
-        closeThreads.add(t);
-      }
+    Collection<InterpreterSetting> intpSettings = interpreterSettings.values();
+    for (final InterpreterSetting intpSetting : intpSettings) {
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Merge them into one line
```
for (InterpreterSetting intpSetting : interpreterSettings.values()) {
```

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.reflect.TypeToken;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: this could be removed

**Comment by zjffdu** [comment]
THanks @jongyoul Just several minor issues, rest look good to me.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java (hunk 2 above)
Comment (jongyoul) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Fixed

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java (hunk 3 above)
Comment (jongyoul) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSettingManager.java]: Fixed

**Comment by jongyoul** [comment]
After finishing CI, I'll merge it

**Comment by jongyoul** [comment]
Test passed. https://travis-ci.org/jongyoul/zeppelin/builds/369961193 Will merge it

**Comment by weand** [comment]
are you going to merge that into branch-0.8 as well?

**Comment by jongyoul** [comment]
@weand Hi, I’d like to do it but need to check the timeline of the release date because the two following PRs should be implemented. Thus I think it would be better to merge it after the rest parts would be implemented. WDYT? @zjffdu when is the target date for the next release?

**Comment by jongyoul** [comment]
@weand Oops. I was confused with another issue. I’ll cherry-pick this into 0.8

**Pull request closed without merging**